### ISSUE TITLE
Problem with closing page on Mx 6.4.1

### ISSUE THREAD

**Issue by JelleDekker** [mendix/IFrame#2]
Using the iFrame widget on a page in Mx 6.4.1 I have encountered the following UI issue:
- the page with the iFrame on it will only close after two clicks. This happens both with the standard "cancel"-button and with an Action-btn triggering a MicroFlow which has a close-page activity.
- After having closed the page with two clicks, it won't reopen anymore. The MF called to open the page is triggered, the "Open page" action is executed, but no page opens
- This happens both on localhost and in the Mendix cloud
- This happens both in Firefox and in Chrome (didn't test other browsers yet)

Furthermore:
- When I remove the widget from the page, I have to manually stop and start the server (localhost). Just pressing F5 after removing the widget from the page leads to a JS error
- After removal of the widget from the page everything works fine again.

Update:
- It's even weirder, after the "open page" that failed, a completely different "open page" now opens the page that didn't want to open before, instead of the page that I wanted to open at that moment.
- It appears that after the "double close", all the new "open page" actions are put in a queue, when I then open a different page, the entire queue is executed
- Problem also occurs on Mx 6.6

**Comment by JelteMX** [comment]
Jelle, could you send a testproject regarding this issue to community [at] mendix.com? We haven't encountered these issues yet, so we'd like to troubleshoot it.

**Comment by JelleDekker** [comment]
I'll send the test project, For now you can already find it at https://iframewidgetbug.mxapps.io/index.html?profile=desktop 
1. Select first item in datagrid (which links to Mendix.com)
2. Click "Edit"
3. Click "Action"
4. Click "Cancel" >> iframe disappears, page stays open
5. Click "Cancel" again >> page closes
6. Click "Action" again >> nothing happens
7. Click "Action" a few more times
8. Click "Cancel" >> the page with the Iframe opens again
9. Click "Cancel" again on the page with iFrame that opened, you'll see that there are more of these opened now (depending on how active you were at step 7)

**Comment by JelleDekker** [comment]
Second scenario:
1. Select first item in datagrid (which links to Mendix.com)
2. Click "Edit"
3. Click "Action"
4. Click "Cancel" twice to close page with iFrame
5. Click "Cancel" again to close the edit page
6. Click "New" >> the page doesn't open, however from the autonumber you can see that the new objects have been created

Third scenario:
1. Select second item in datagrid (which links to Google.nl)
2. Click "Edit"
3. Click "Action"
4. Click "Cancel" >> all works fine when the iFrame was not able to load it's content

**Comment by JelteMX** [comment]
Jelle, sorry, could you send the invitation to jelte.lagendijk [at] mendix.com? Community is a shared mailbox, not a Mendix account :-)

**Comment by edwinmp** [comment]
Hi,

Investigated this problem and in my test project, it's originating from two Mendix App Cloud widgets: Mendix App Switcher and Mendix Profile Menu. These two widgets appear to be conflicting with the IFrame widget.

The code below from the Profile Menu & App Switcher widgets raise a multipleDefine(…) dojo error because it appends the same scripts multiple times.

```
           ioScript.get({
                url: url + 'mendixtoolbar/js/buttonservices.js?PP_6.20',
                error: dojo.hitch(this, function (e) {
                    console && console.log('Mendix AppSwitcher/ProfileMenu could not load external script: ', e);
                })
            });
```

On cancel, another error is triggered, originating from the same buttonservice script that the two widgets are appending:

> Failed to execute 'postMessage' on 'DOMWindow': The target origin provided ('https://home.mendix.com') does not match the recipient window's origin ('http://localhost:8080').

This is because the AppSwitcher and ProfileMenu widgets, which generate iframes of their own, are attempting to access an external source running over https yet their origin is not secure (http).

Well, this was all on the assumption that you have these two widgets active in your project too. If so, then I suggest you decide how much you need them in there and consider removing them.

**Comment by JelleDekker** [comment]
I don't have those widgets in the project where this error originated I'm afraid. Also in the project where I first noticed this, the browser console doesn't show any errors.

Edit:
I just updated the example project to resemble the real one more. Removed the two widgets you mentioned and instead added the FormatString widget, which we do use.

**Comment by JelteMX** [comment]
I checked the testproject you provided (thanks by the way 😄 ) and found the issue. The reason for this issue is actually that when you update the context object (actually an update is triggered as well when you press Cancel), the url was reloaded. So I changed some of the code, in which that doesn't happen anymore.

Please note (this will be noted in the documentation as well) that some websites do fail to load in an iframe due to restrictions, but that it's pretty hard to determine when this happens. For a website like google.nl, it is blocked in the browser which is impossible circumvent. If for example I load http://nu.nl these problems do not arise.

I'll have a fix for you to test

**Comment by JelteMX** [comment]
Please test [this version](https://github.com/mendix/IFrame/raw/master/test/widgets/IFrame.mpk) and provide me with some feedback whether or not this issue is fixed for you

**Comment by edwinmp** [comment]
The [CKEditorForMendix](https://github.com/mendix/CKEditorForMendix) would be an excellent alternative. it has a pretty good iFrame plugin with similar options and a few extra.

**Comment by JelleDekker** [comment]
@JelteMX The solution seems to work. However in my original application the problem still occurred, I managed to fix it there by removing a "refresh" in the action that processed the users input in the screen.

I tried to reproduce the situation which still gave the problems also in the test project, but there everything goes fine when I do.

**Comment by JelteMX** [comment]
Great! I'll create an update for the AppStore

**Comment by JelteMX** [comment]
Fixed in [3.1.1](https://github.com/mendix/IFrame/releases/tag/v3.1.1)
